From dd2f359c4656bd457b691986f2ba6a44afee3299 Mon Sep 17 00:00:00 2001
From: Arnold Wang <Arnold.Wang@ingeniux.com>
Date: Thu, 14 Dec 2023 10:28:27 -0800
Subject: [PATCH 1/8] Fix bug of cheerio loader eventually failing with not
 iterable error Added more console logging for web crawl to better indicating
 progress

---
 .../nodes/documentloaders/Cheerio/Cheerio.ts  | 22 +++--
 packages/components/src/utils.ts              | 12 +--
 packages/server/src/index.ts                  | 89 +++++++++----------
 3 files changed, 68 insertions(+), 55 deletions(-)

diff --git a/packages/components/nodes/documentloaders/Cheerio/Cheerio.ts b/packages/components/nodes/documentloaders/Cheerio/Cheerio.ts
index aa899bcbe4c..c06b736f43c 100644
--- a/packages/components/nodes/documentloaders/Cheerio/Cheerio.ts
+++ b/packages/components/nodes/documentloaders/Cheerio/Cheerio.ts
@@ -1,10 +1,10 @@
-import { INode, INodeData, INodeParams } from '../../../src/Interface'
-import { TextSplitter } from 'langchain/text_splitter'
+import { SelectorType } from 'cheerio'
+import { parse } from 'css-what'
 import { CheerioWebBaseLoader, WebBaseLoaderParams } from 'langchain/document_loaders/web/cheerio'
+import { TextSplitter } from 'langchain/text_splitter'
 import { test } from 'linkifyjs'
-import { parse } from 'css-what'
 import { webCrawl, xmlScrape } from '../../../src'
-import { SelectorType } from 'cheerio'
+import { INode, INodeData, INodeParams } from '../../../src/Interface'
 
 class Cheerio_DocumentLoaders implements INode {
     label: string
@@ -110,11 +110,13 @@ class Cheerio_DocumentLoaders implements INode {
             try {
                 let docs = []
                 const loader = new CheerioWebBaseLoader(url, params)
+                console.info(`scraping - loading url ${url}`)
                 if (textSplitter) {
                     docs = await loader.loadAndSplit(textSplitter)
                 } else {
                     docs = await loader.load()
                 }
+                console.info(`scraping - loaded ${docs.length} docs from ${url}`)
                 return docs
             } catch (err) {
                 if (process.env.DEBUG === 'true') console.error(`error in CheerioWebBaseLoader: ${err.message}, on page: ${url}`)
@@ -126,10 +128,17 @@ class Cheerio_DocumentLoaders implements INode {
             if (process.env.DEBUG === 'true') console.info(`Start ${relativeLinksMethod}`)
             if (!limit) limit = '10'
             else if (parseInt(limit) < 0) throw new Error('Limit cannot be less than 0')
+            console.info(`scrape limit: ${limit}`)
+            console.info(`scraping url: ${url}`)
             const pages: string[] =
                 relativeLinksMethod === 'webCrawl' ? await webCrawl(url, parseInt(limit)) : await xmlScrape(url, parseInt(limit))
+
             if (process.env.DEBUG === 'true') console.info(`pages: ${JSON.stringify(pages)}, length: ${pages.length}`)
-            if (!pages || pages.length === 0) throw new Error('No relative links found')
+            if (!pages || !Array.isArray(pages) || pages.length === 0) {
+                console.error(`No relative links found for ${url}`)
+                return
+            }
+            console.info(`scraping found ${pages.length} pages: ${pages.join(', ')}`)
             for (const page of pages) {
                 docs.push(...(await cheerioLoader(page)))
             }
@@ -141,6 +150,7 @@ class Cheerio_DocumentLoaders implements INode {
         if (metadata) {
             const parsedMetadata = typeof metadata === 'object' ? metadata : JSON.parse(metadata)
             let finaldocs = []
+
             for (const doc of docs) {
                 const newdoc = {
                     ...doc,
@@ -154,6 +164,8 @@ class Cheerio_DocumentLoaders implements INode {
             return finaldocs
         }
 
+        console.info(`scraped ${docs.length} docs from ${url}`)
+
         return docs
     }
 }
diff --git a/packages/components/src/utils.ts b/packages/components/src/utils.ts
index 404f7c75d8f..90e4c2110d4 100644
--- a/packages/components/src/utils.ts
+++ b/packages/components/src/utils.ts
@@ -1,14 +1,14 @@
 import axios from 'axios'
 import { load } from 'cheerio'
+import { AES, enc } from 'crypto-js'
 import * as fs from 'fs'
-import * as path from 'path'
 import { JSDOM } from 'jsdom'
-import { z } from 'zod'
-import { DataSource } from 'typeorm'
-import { ICommonObject, IDatabaseEntity, IMessage, INodeData } from './Interface'
-import { AES, enc } from 'crypto-js'
 import { ChatMessageHistory } from 'langchain/memory'
 import { AIMessage, HumanMessage } from 'langchain/schema'
+import * as path from 'path'
+import { DataSource } from 'typeorm'
+import { z } from 'zod'
+import { ICommonObject, IDatabaseEntity, IMessage, INodeData } from './Interface'
 
 export const numberOrExpressionRegex = '^(\\d+\\.?\\d*|{{.*}})$' //return true if string consists only numbers OR expression {{}}
 export const notEmptyRegex = '(.|\\s)*\\S(.|\\s)*' //return true if string is not empty or blank
@@ -278,6 +278,7 @@ async function crawl(baseURL: string, currentURL: string, pages: string[], limit
 
     if (process.env.DEBUG === 'true') console.info(`actively crawling ${currentURL}`)
     try {
+        console.info(`crawling ${currentURL}`)
         const resp = await fetch(currentURL)
 
         if (resp.status > 399) {
@@ -292,6 +293,7 @@ async function crawl(baseURL: string, currentURL: string, pages: string[], limit
         }
 
         const htmlBody = await resp.text()
+        console.info(`crawled ${currentURL}`)
         const nextURLs = getURLsFromHTML(htmlBody, baseURL)
         for (const nextURL of nextURLs) {
             pages = await crawl(baseURL, nextURL, pages, limit)
diff --git a/packages/server/src/index.ts b/packages/server/src/index.ts
index fb4a5f5a0cf..fbd9816c085 100644
--- a/packages/server/src/index.ts
+++ b/packages/server/src/index.ts
@@ -1,67 +1,66 @@
+import axios from 'axios'
+import cors from 'cors'
 import express, { NextFunction, Request, Response } from 'express'
+import basicAuth from 'express-basic-auth'
+import { ICommonObject, IMessage, INodeOptionsValue } from 'flowise-components'
+import * as fs from 'fs'
+import http from 'http'
+import { Client } from 'langchainhub'
+import { cloneDeep, isEqual, omit, uniqWith } from 'lodash'
 import multer from 'multer'
+import OpenAI from 'openai'
 import path from 'path'
-import cors from 'cors'
-import http from 'http'
-import * as fs from 'fs'
-import basicAuth from 'express-basic-auth'
 import { Server } from 'socket.io'
-import logger from './utils/logger'
-import { expressRequestLogger } from './utils/logger'
+import { Between, FindOptionsWhere, IsNull } from 'typeorm'
 import { v4 as uuidv4 } from 'uuid'
-import OpenAI from 'openai'
-import { Between, IsNull, FindOptionsWhere } from 'typeorm'
+import { CachePool } from './CachePool'
+import { ChatflowPool } from './ChatflowPool'
+import { getDataSource } from './DataSource'
 import {
     IChatFlow,
-    IncomingInput,
+    IChatMessage,
+    ICredentialReturnResponse,
+    INodeData,
+    IReactFlowEdge,
     IReactFlowNode,
     IReactFlowObject,
-    INodeData,
-    ICredentialReturnResponse,
-    chatType,
-    IChatMessage,
-    IReactFlowEdge
+    IncomingInput,
+    chatType
 } from './Interface'
+import { NodesPool } from './NodesPool'
+import { Assistant } from './database/entities/Assistant'
+import { ChatFlow } from './database/entities/ChatFlow'
+import { ChatMessage } from './database/entities/ChatMessage'
+import { Credential } from './database/entities/Credential'
+import { Tool } from './database/entities/Tool'
 import {
-    getNodeModulesPackagePath,
-    getStartingNodes,
     buildLangchain,
-    getEndingNode,
+    checkMemorySessionId,
+    clearAllSessionMemory,
+    clearSessionMemoryFromViewMessageDialog,
     constructGraphs,
-    resolveVariables,
-    isStartNodeDependOnInput,
-    mapMimeTypeToInputField,
-    findAvailableConfigs,
-    isSameOverrideConfig,
-    isFlowValidForStream,
     databaseEntities,
-    transformToCredentialEntity,
     decryptCredentialData,
-    clearAllSessionMemory,
-    replaceInputsWithConfig,
+    findAvailableConfigs,
     getEncryptionKey,
-    checkMemorySessionId,
-    clearSessionMemoryFromViewMessageDialog,
+    getEndingNode,
+    getNodeModulesPackagePath,
+    getStartingNodes,
     getUserHome,
-    replaceChatHistory
+    isFlowValidForStream,
+    isSameOverrideConfig,
+    isStartNodeDependOnInput,
+    mapMimeTypeToInputField,
+    replaceChatHistory,
+    replaceInputsWithConfig,
+    resolveVariables,
+    transformToCredentialEntity
 } from './utils'
-import { cloneDeep, omit, uniqWith, isEqual } from 'lodash'
-import { getDataSource } from './DataSource'
-import { NodesPool } from './NodesPool'
-import { ChatFlow } from './database/entities/ChatFlow'
-import { ChatMessage } from './database/entities/ChatMessage'
-import { Credential } from './database/entities/Credential'
-import { Tool } from './database/entities/Tool'
-import { Assistant } from './database/entities/Assistant'
-import { ChatflowPool } from './ChatflowPool'
-import { CachePool } from './CachePool'
-import { ICommonObject, IMessage, INodeOptionsValue } from 'flowise-components'
-import { createRateLimiter, getRateLimiter, initializeRateLimiter } from './utils/rateLimit'
-import { addAPIKey, compareKeys, deleteAPIKey, getApiKey, getAPIKeys, updateAPIKey } from './utils/apiKey'
 import { sanitizeMiddleware } from './utils/XSS'
-import axios from 'axios'
-import { Client } from 'langchainhub'
+import { addAPIKey, compareKeys, deleteAPIKey, getAPIKeys, getApiKey, updateAPIKey } from './utils/apiKey'
 import { parsePrompt } from './utils/hub'
+import logger, { expressRequestLogger } from './utils/logger'
+import { createRateLimiter, getRateLimiter, initializeRateLimiter } from './utils/rateLimit'
 
 export class App {
     app: express.Application
@@ -1647,7 +1646,7 @@ export async function getAllChatFlow(): Promise<IChatFlow[]> {
 export async function start(): Promise<void> {
     serverApp = new App()
 
-    const port = parseInt(process.env.PORT || '', 10) || 3000
+    const port = parseInt(process.env.PORT || '', 10) || 3333
     const server = http.createServer(serverApp.app)
 
     const io = new Server(server, {

From 10f850a33289f5e344c93c6bf84b0ac4d13b5719 Mon Sep 17 00:00:00 2001
From: Arnold Wang <Arnold.Wang@ingeniux.com>
Date: Thu, 14 Dec 2023 16:55:00 -0800
Subject: [PATCH 2/8] exclude archives upfront to prevent time out and not
 interable error

---
 packages/components/src/utils.ts | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/packages/components/src/utils.ts b/packages/components/src/utils.ts
index 90e4c2110d4..c37976b4f53 100644
--- a/packages/components/src/utils.ts
+++ b/packages/components/src/utils.ts
@@ -252,6 +252,7 @@ function normalizeURL(urlString: string): string {
     }
     return hostPath
 }
+const LARGE_FILE_EXTENSIONS = ['zip', 'tar', 'rar', 'jar', 'arj', 'gz'] //todo: add full listing
 
 /**
  * Recursive crawl using normalizeURL and getURLsFromHTML
@@ -270,6 +271,14 @@ async function crawl(baseURL: string, currentURL: string, pages: string[], limit
     if (baseURLObj.hostname !== currentURLObj.hostname) return pages
 
     const normalizeCurrentURL = baseURLObj.protocol + '//' + normalizeURL(currentURL)
+
+    const lastSec = normalizeCurrentURL.substring(normalizeCurrentURL.lastIndexOf('/') + 1)
+    const dotPos = lastSec.lastIndexOf('.')
+    const urlExt = dotPos > -1 && dotPos < lastSec.length - 1 ? lastSec.substring(dotPos + 1) : ''
+
+    //fix issue with interable error when crawing a zip file, most likely timed out
+    if (!!urlExt && LARGE_FILE_EXTENSIONS.includes(urlExt)) return pages
+
     if (pages.includes(normalizeCurrentURL)) {
         return pages
     }

From 9ad89b8451800996f641e735d038e80d9718ba5e Mon Sep 17 00:00:00 2001
From: Arnold Wang <Arnold.Wang@ingeniux.com>
Date: Sun, 17 Dec 2023 18:34:07 -0800
Subject: [PATCH 3/8] Fix issue of bypassing cheero crawl limit due to
 incorrect logic

---
 packages/components/src/utils.ts | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/packages/components/src/utils.ts b/packages/components/src/utils.ts
index c37976b4f53..a7b602facbb 100644
--- a/packages/components/src/utils.ts
+++ b/packages/components/src/utils.ts
@@ -266,7 +266,10 @@ async function crawl(baseURL: string, currentURL: string, pages: string[], limit
     const baseURLObj = new URL(baseURL)
     const currentURLObj = new URL(currentURL)
 
-    if (limit !== 0 && pages.length === limit) return pages
+    if (limit > 0 && pages.length >= limit) {
+        console.info(`crawl limit reached: ${limit}`)
+        return pages
+    }
 
     if (baseURLObj.hostname !== currentURLObj.hostname) return pages
 

From ce307f8296a4137d79fd3e7f846034cdf0befe7f Mon Sep 17 00:00:00 2001
From: Arnold Wang <Arnold.Wang@ingeniux.com>
Date: Mon, 18 Dec 2023 10:17:43 -0800
Subject: [PATCH 4/8] Fix cheerio limit not controlling actual documents to
 upsert issue

---
 .../nodes/documentloaders/Cheerio/Cheerio.ts  | 27 ++++++++++++++-----
 .../nodes/vectorstores/Pinecone/Pinecone.ts   | 13 +++------
 packages/components/src/utils.ts              |  3 +++
 3 files changed, 28 insertions(+), 15 deletions(-)

diff --git a/packages/components/nodes/documentloaders/Cheerio/Cheerio.ts b/packages/components/nodes/documentloaders/Cheerio/Cheerio.ts
index c06b736f43c..092788be300 100644
--- a/packages/components/nodes/documentloaders/Cheerio/Cheerio.ts
+++ b/packages/components/nodes/documentloaders/Cheerio/Cheerio.ts
@@ -119,7 +119,7 @@ class Cheerio_DocumentLoaders implements INode {
                 console.info(`scraping - loaded ${docs.length} docs from ${url}`)
                 return docs
             } catch (err) {
-                if (process.env.DEBUG === 'true') console.error(`error in CheerioWebBaseLoader: ${err.message}, on page: ${url}`)
+                console.error(`error in CheerioWebBaseLoader: ${err.message}, on page: ${url}`)
             }
         }
 
@@ -130,18 +130,29 @@ class Cheerio_DocumentLoaders implements INode {
             else if (parseInt(limit) < 0) throw new Error('Limit cannot be less than 0')
             console.info(`scrape limit: ${limit}`)
             console.info(`scraping url: ${url}`)
-            const pages: string[] =
+            let pages: string[] =
                 relativeLinksMethod === 'webCrawl' ? await webCrawl(url, parseInt(limit)) : await xmlScrape(url, parseInt(limit))
 
             if (process.env.DEBUG === 'true') console.info(`pages: ${JSON.stringify(pages)}, length: ${pages.length}`)
             if (!pages || !Array.isArray(pages) || pages.length === 0) {
-                console.error(`No relative links found for ${url}`)
+                console.warn(`No relative links found for ${url}`)
                 return
             }
-            console.info(`scraping found ${pages.length} pages: ${pages.join(', ')}`)
-            for (const page of pages) {
-                docs.push(...(await cheerioLoader(page)))
+
+            if (!!limit && parseInt(limit) > 0) {
+                console.info(`scraping limit to ${limit}`)
+                pages = pages.slice(0, parseInt(limit)) // limit docs to be returned
+            }
+
+            try {
+                console.info(`scraping found ${pages.length} pages: ${pages.join(', ')}`)
+                for (const page of pages) {
+                    docs.push(...(await cheerioLoader(page)))
+                }
+            } catch (err) {
+                console.error(`error in CheerioWebBaseLoader: ${err.message}, on page: ${url}`)
             }
+
             if (process.env.DEBUG === 'true') console.info(`Finish ${relativeLinksMethod}`)
         } else {
             docs = await cheerioLoader(url)
@@ -165,6 +176,10 @@ class Cheerio_DocumentLoaders implements INode {
         }
 
         console.info(`scraped ${docs.length} docs from ${url}`)
+        if (!!limit && parseInt(limit) > 0 && docs.length > parseInt(limit)) {
+            console.info(`scraped docs limiting to ${limit}`)
+            docs = docs.slice(0, parseInt(limit)) // limit docs to be returned
+        }
 
         return docs
     }
diff --git a/packages/components/nodes/vectorstores/Pinecone/Pinecone.ts b/packages/components/nodes/vectorstores/Pinecone/Pinecone.ts
index 4ece4720a35..d9ad123aab1 100644
--- a/packages/components/nodes/vectorstores/Pinecone/Pinecone.ts
+++ b/packages/components/nodes/vectorstores/Pinecone/Pinecone.ts
@@ -1,8 +1,8 @@
-import { flatten } from 'lodash'
 import { Pinecone } from '@pinecone-database/pinecone'
-import { PineconeLibArgs, PineconeStore } from 'langchain/vectorstores/pinecone'
-import { Embeddings } from 'langchain/embeddings/base'
 import { Document } from 'langchain/document'
+import { Embeddings } from 'langchain/embeddings/base'
+import { PineconeLibArgs, PineconeStore } from 'langchain/vectorstores/pinecone'
+import { flatten } from 'lodash'
 import { ICommonObject, INode, INodeData, INodeOutputsValue, INodeParams } from '../../../src/Interface'
 import { getBaseClasses, getCredentialData, getCredentialParam } from '../../../src/utils'
 
@@ -113,12 +113,7 @@ class Pinecone_VectorStores implements INode {
             const pineconeIndex = client.Index(index)
 
             const flattenDocs = docs && docs.length ? flatten(docs) : []
-            const finalDocs = []
-            for (let i = 0; i < flattenDocs.length; i += 1) {
-                if (flattenDocs[i] && flattenDocs[i].pageContent) {
-                    finalDocs.push(new Document(flattenDocs[i]))
-                }
-            }
+            const finalDocs = flattenDocs.filter((doc) => !!doc?.pageContent).map((doc) => new Document(doc))
 
             const obj: PineconeLibArgs = {
                 pineconeIndex
diff --git a/packages/components/src/utils.ts b/packages/components/src/utils.ts
index a7b602facbb..563ded69121 100644
--- a/packages/components/src/utils.ts
+++ b/packages/components/src/utils.ts
@@ -313,6 +313,9 @@ async function crawl(baseURL: string, currentURL: string, pages: string[], limit
     } catch (err) {
         if (process.env.DEBUG === 'true') console.error(`error in fetch url: ${err.message}, on page: ${currentURL}`)
     }
+
+    console.info(`crawled ${pages.length} pages so far, limit: ${limit}}`)
+
     return pages
 }
 

From 02d0da9e96eed680f15d4cf9432a070813cc25ed Mon Sep 17 00:00:00 2001
From: Arnold Wang <Arnold.Wang@ingeniux.com>
Date: Tue, 19 Dec 2023 15:05:11 -0800
Subject: [PATCH 5/8] Added inclusion and exclusion settings for web crawl.
 Allowing finer control of web crawling

---
 .../nodes/documentloaders/Cheerio/Cheerio.ts  | 42 +++++++++++++++++--
 packages/components/src/utils.ts              | 37 ++++++++++++----
 2 files changed, 68 insertions(+), 11 deletions(-)

diff --git a/packages/components/nodes/documentloaders/Cheerio/Cheerio.ts b/packages/components/nodes/documentloaders/Cheerio/Cheerio.ts
index 092788be300..958622a90e6 100644
--- a/packages/components/nodes/documentloaders/Cheerio/Cheerio.ts
+++ b/packages/components/nodes/documentloaders/Cheerio/Cheerio.ts
@@ -76,6 +76,23 @@ class Cheerio_DocumentLoaders implements INode {
                 optional: true,
                 additionalParams: true
             },
+            {
+                label: 'Base URL Prefixes',
+                name: 'urlFilter',
+                type: 'string',
+                description: 'Delimited by comma. If specified, only links that start with this URL will be retrieved. (Web Crawl only)',
+                optional: true,
+                additionalParams: true
+            },
+            {
+                label: 'Excluded URL Prefixes',
+                name: 'exUrlFilter',
+                type: 'string',
+                description:
+                    "Delimited by comma. If specified, only links that don't start with this URL will be retrieved. (Web Crawl only)",
+                optional: true,
+                additionalParams: true
+            },
             {
                 label: 'Metadata',
                 name: 'metadata',
@@ -106,9 +123,25 @@ class Cheerio_DocumentLoaders implements INode {
             params['selector'] = selector
         }
 
-        async function cheerioLoader(url: string): Promise<any> {
+        const baseUrlFilters = ((nodeData.inputs?.urlFilter as string)?.trim()?.toLowerCase() || '').split(',').filter((x) => !!x)
+        const exBaseUrlFilter = ((nodeData.inputs?.exUrlFilter as string)?.trim()?.toLowerCase() || '').split(',').filter((x) => !!x)
+
+        console.info(`Prefix urls: ${baseUrlFilters.join(',')}`)
+        console.info(`Excluded urls: ${exBaseUrlFilter.join(',')}`)
+
+        async function cheerioLoader(url: string): Promise<any[]> {
             try {
-                let docs = []
+                let docs = [] as any[]
+                if (!!baseUrlFilters && !baseUrlFilters.some((baseUrl) => url.toLowerCase().startsWith(baseUrl))) {
+                    console.info(`scraping - skipping url ${url} because it does not start with ${baseUrlFilters}`)
+                    return docs
+                }
+
+                if (!!exBaseUrlFilter && exBaseUrlFilter.some((exBaseUrl) => url.toLowerCase().startsWith(exBaseUrl))) {
+                    console.info(`scraping - skipping url ${url} because it starts with ${exBaseUrlFilter}`)
+                    return docs
+                }
+
                 const loader = new CheerioWebBaseLoader(url, params)
                 console.info(`scraping - loading url ${url}`)
                 if (textSplitter) {
@@ -120,6 +153,7 @@ class Cheerio_DocumentLoaders implements INode {
                 return docs
             } catch (err) {
                 console.error(`error in CheerioWebBaseLoader: ${err.message}, on page: ${url}`)
+                return []
             }
         }
 
@@ -131,7 +165,9 @@ class Cheerio_DocumentLoaders implements INode {
             console.info(`scrape limit: ${limit}`)
             console.info(`scraping url: ${url}`)
             let pages: string[] =
-                relativeLinksMethod === 'webCrawl' ? await webCrawl(url, parseInt(limit)) : await xmlScrape(url, parseInt(limit))
+                relativeLinksMethod === 'webCrawl'
+                    ? await webCrawl(url, parseInt(limit), baseUrlFilters, exBaseUrlFilter)
+                    : await xmlScrape(url, parseInt(limit))
 
             if (process.env.DEBUG === 'true') console.info(`pages: ${JSON.stringify(pages)}, length: ${pages.length}`)
             if (!pages || !Array.isArray(pages) || pages.length === 0) {
diff --git a/packages/components/src/utils.ts b/packages/components/src/utils.ts
index 563ded69121..5b3218dbcf2 100644
--- a/packages/components/src/utils.ts
+++ b/packages/components/src/utils.ts
@@ -243,13 +243,18 @@ function getURLsFromHTML(htmlBody: string, baseURL: string): string[] {
  * @param {string} urlString
  * @returns {string}
  */
-function normalizeURL(urlString: string): string {
+function normalizeURL(urlString: string, removeBookmark?: boolean): string {
     const urlObj = new URL(urlString)
-    const hostPath = urlObj.hostname + urlObj.pathname
+    let hostPath = urlObj.hostname + urlObj.pathname
     if (hostPath.length > 0 && hostPath.slice(-1) == '/') {
         // handling trailing slash
         return hostPath.slice(0, -1)
     }
+    if (removeBookmark && urlString.includes('#')) {
+        // handling bookmark
+        hostPath = hostPath.substring(0, hostPath.indexOf('#'))
+    }
+
     return hostPath
 }
 const LARGE_FILE_EXTENSIONS = ['zip', 'tar', 'rar', 'jar', 'arj', 'gz'] //todo: add full listing
@@ -262,7 +267,14 @@ const LARGE_FILE_EXTENSIONS = ['zip', 'tar', 'rar', 'jar', 'arj', 'gz'] //todo:
  * @param {number} limit
  * @returns {Promise<string[]>}
  */
-async function crawl(baseURL: string, currentURL: string, pages: string[], limit: number): Promise<string[]> {
+async function crawl(
+    baseURL: string,
+    currentURL: string,
+    pages: string[],
+    limit: number,
+    prefixUrls?: string[],
+    exPrefixUrls?: string[]
+): Promise<string[]> {
     const baseURLObj = new URL(baseURL)
     const currentURLObj = new URL(currentURL)
 
@@ -273,7 +285,7 @@ async function crawl(baseURL: string, currentURL: string, pages: string[], limit
 
     if (baseURLObj.hostname !== currentURLObj.hostname) return pages
 
-    const normalizeCurrentURL = baseURLObj.protocol + '//' + normalizeURL(currentURL)
+    const normalizeCurrentURL = baseURLObj.protocol + '//' + normalizeURL(currentURL, true)
 
     const lastSec = normalizeCurrentURL.substring(normalizeCurrentURL.lastIndexOf('/') + 1)
     const dotPos = lastSec.lastIndexOf('.')
@@ -282,13 +294,22 @@ async function crawl(baseURL: string, currentURL: string, pages: string[], limit
     //fix issue with interable error when crawing a zip file, most likely timed out
     if (!!urlExt && LARGE_FILE_EXTENSIONS.includes(urlExt)) return pages
 
+    if (!!prefixUrls && !prefixUrls.some((prefixUrl) => normalizeCurrentURL.toLowerCase().startsWith(prefixUrl))) {
+        console.info(`skipping url ${normalizeCurrentURL} because it does not start with any required prefix urls.`)
+        return pages
+    }
+
+    if (!!exPrefixUrls && exPrefixUrls.some((exPrefixUrl) => normalizeCurrentURL.toLowerCase().startsWith(exPrefixUrl))) {
+        console.info(`skipping url ${normalizeCurrentURL} because it starts with one or more excluded prefix urls.`)
+        return pages
+    }
+
     if (pages.includes(normalizeCurrentURL)) {
         return pages
     }
 
     pages.push(normalizeCurrentURL)
 
-    if (process.env.DEBUG === 'true') console.info(`actively crawling ${currentURL}`)
     try {
         console.info(`crawling ${currentURL}`)
         const resp = await fetch(currentURL)
@@ -311,7 +332,7 @@ async function crawl(baseURL: string, currentURL: string, pages: string[], limit
             pages = await crawl(baseURL, nextURL, pages, limit)
         }
     } catch (err) {
-        if (process.env.DEBUG === 'true') console.error(`error in fetch url: ${err.message}, on page: ${currentURL}`)
+        console.error(`error in fetch url: ${err.message}, on page: ${currentURL}`)
     }
 
     console.info(`crawled ${pages.length} pages so far, limit: ${limit}}`)
@@ -325,10 +346,10 @@ async function crawl(baseURL: string, currentURL: string, pages: string[], limit
  * @param {number} limit
  * @returns {Promise<string[]>}
  */
-export async function webCrawl(stringURL: string, limit: number): Promise<string[]> {
+export async function webCrawl(stringURL: string, limit: number, baseUrls?: string[], exBaseUrls?: string[]): Promise<string[]> {
     const URLObj = new URL(stringURL)
     const modifyURL = stringURL.slice(-1) === '/' ? stringURL.slice(0, -1) : stringURL
-    return await crawl(URLObj.protocol + '//' + URLObj.hostname, modifyURL, [], limit)
+    return await crawl(URLObj.protocol + '//' + URLObj.hostname, modifyURL, [], limit, baseUrls, exBaseUrls)
 }
 
 export function getURLsFromXML(xmlBody: string, limit: number): string[] {

From 86bc78ac6e21a09e243ff337c1ea4a96150c5923 Mon Sep 17 00:00:00 2001
From: Arnold Wang <Arnold.Wang@ingeniux.com>
Date: Tue, 19 Dec 2023 16:07:55 -0800
Subject: [PATCH 6/8] Fix cheerio for not filtering out loaded docs and
 excluded doc, causing dupes and not valid docs

---
 .../nodes/documentloaders/Cheerio/Cheerio.ts  | 25 ++++++++++++++++---
 1 file changed, 21 insertions(+), 4 deletions(-)

diff --git a/packages/components/nodes/documentloaders/Cheerio/Cheerio.ts b/packages/components/nodes/documentloaders/Cheerio/Cheerio.ts
index 958622a90e6..6488d86bf31 100644
--- a/packages/components/nodes/documentloaders/Cheerio/Cheerio.ts
+++ b/packages/components/nodes/documentloaders/Cheerio/Cheerio.ts
@@ -1,5 +1,6 @@
 import { SelectorType } from 'cheerio'
 import { parse } from 'css-what'
+import { Document } from 'langchain/document'
 import { CheerioWebBaseLoader, WebBaseLoaderParams } from 'langchain/document_loaders/web/cheerio'
 import { TextSplitter } from 'langchain/text_splitter'
 import { test } from 'linkifyjs'
@@ -129,9 +130,11 @@ class Cheerio_DocumentLoaders implements INode {
         console.info(`Prefix urls: ${baseUrlFilters.join(',')}`)
         console.info(`Excluded urls: ${exBaseUrlFilter.join(',')}`)
 
-        async function cheerioLoader(url: string): Promise<any[]> {
+        const loadedDocUrls = new Set<string>()
+
+        async function cheerioLoader(url: string): Promise<Document[]> {
             try {
-                let docs = [] as any[]
+                let docs = [] as Document[]
                 if (!!baseUrlFilters && !baseUrlFilters.some((baseUrl) => url.toLowerCase().startsWith(baseUrl))) {
                     console.info(`scraping - skipping url ${url} because it does not start with ${baseUrlFilters}`)
                     return docs
@@ -149,8 +152,22 @@ class Cheerio_DocumentLoaders implements INode {
                 } else {
                     docs = await loader.load()
                 }
-                console.info(`scraping - loaded ${docs.length} docs from ${url}`)
-                return docs
+
+                const newDocs = docs.filter(
+                    (doc) =>
+                        !loadedDocUrls.has(doc.metadata.source) &&
+                        (!baseUrlFilters || baseUrlFilters.some((baseUrl) => doc.metadata.source.toLowerCase().startsWith(baseUrl))) &&
+                        (!exBaseUrlFilter || !exBaseUrlFilter.some((exBaseUrl) => doc.metadata.source.toLowerCase().startsWith(exBaseUrl)))
+                )
+
+                newDocs
+                    .map((doc) => doc.metadata.source)
+                    .forEach((docUrl) => {
+                        if (!loadedDocUrls.has(docUrl)) loadedDocUrls.add(docUrl)
+                    })
+
+                console.info(`scraping - loaded ${docs.length} new docs from ${url}`)
+                return newDocs
             } catch (err) {
                 console.error(`error in CheerioWebBaseLoader: ${err.message}, on page: ${url}`)
                 return []

From 4db5ac3de7ba15bf3004ff9fd2ee60b03fee938c Mon Sep 17 00:00:00 2001
From: Arnold Wang <Arnold.Wang@ingeniux.com>
Date: Wed, 20 Dec 2023 09:38:34 -0800
Subject: [PATCH 7/8] prefix and ex prefix url checking to handle empty listing
 scenario

---
 packages/components/nodes/documentloaders/Cheerio/Cheerio.ts | 4 ++--
 packages/components/src/utils.ts                             | 4 ++--
 2 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/packages/components/nodes/documentloaders/Cheerio/Cheerio.ts b/packages/components/nodes/documentloaders/Cheerio/Cheerio.ts
index 6488d86bf31..fc789445f87 100644
--- a/packages/components/nodes/documentloaders/Cheerio/Cheerio.ts
+++ b/packages/components/nodes/documentloaders/Cheerio/Cheerio.ts
@@ -135,12 +135,12 @@ class Cheerio_DocumentLoaders implements INode {
         async function cheerioLoader(url: string): Promise<Document[]> {
             try {
                 let docs = [] as Document[]
-                if (!!baseUrlFilters && !baseUrlFilters.some((baseUrl) => url.toLowerCase().startsWith(baseUrl))) {
+                if (!!baseUrlFilters?.length && !baseUrlFilters.some((baseUrl) => url.toLowerCase().startsWith(baseUrl))) {
                     console.info(`scraping - skipping url ${url} because it does not start with ${baseUrlFilters}`)
                     return docs
                 }
 
-                if (!!exBaseUrlFilter && exBaseUrlFilter.some((exBaseUrl) => url.toLowerCase().startsWith(exBaseUrl))) {
+                if (!!exBaseUrlFilter?.length && exBaseUrlFilter.some((exBaseUrl) => url.toLowerCase().startsWith(exBaseUrl))) {
                     console.info(`scraping - skipping url ${url} because it starts with ${exBaseUrlFilter}`)
                     return docs
                 }
diff --git a/packages/components/src/utils.ts b/packages/components/src/utils.ts
index 5b3218dbcf2..e27652bdfe2 100644
--- a/packages/components/src/utils.ts
+++ b/packages/components/src/utils.ts
@@ -294,12 +294,12 @@ async function crawl(
     //fix issue with interable error when crawing a zip file, most likely timed out
     if (!!urlExt && LARGE_FILE_EXTENSIONS.includes(urlExt)) return pages
 
-    if (!!prefixUrls && !prefixUrls.some((prefixUrl) => normalizeCurrentURL.toLowerCase().startsWith(prefixUrl))) {
+    if (!!prefixUrls?.length && !prefixUrls.some((prefixUrl) => normalizeCurrentURL.toLowerCase().startsWith(prefixUrl))) {
         console.info(`skipping url ${normalizeCurrentURL} because it does not start with any required prefix urls.`)
         return pages
     }
 
-    if (!!exPrefixUrls && exPrefixUrls.some((exPrefixUrl) => normalizeCurrentURL.toLowerCase().startsWith(exPrefixUrl))) {
+    if (!!exPrefixUrls?.length && exPrefixUrls.some((exPrefixUrl) => normalizeCurrentURL.toLowerCase().startsWith(exPrefixUrl))) {
         console.info(`skipping url ${normalizeCurrentURL} because it starts with one or more excluded prefix urls.`)
         return pages
     }

From fd3bf9c6e8de5695492c6646179183c0f091f4d3 Mon Sep 17 00:00:00 2001
From: Arnold Wang <Arnold.Wang@ingeniux.com>
Date: Wed, 20 Dec 2023 09:41:06 -0800
Subject: [PATCH 8/8] change port back to 3000

---
 packages/server/src/index.ts | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/packages/server/src/index.ts b/packages/server/src/index.ts
index fbd9816c085..4f9c61e4d5a 100644
--- a/packages/server/src/index.ts
+++ b/packages/server/src/index.ts
@@ -1646,7 +1646,7 @@ export async function getAllChatFlow(): Promise<IChatFlow[]> {
 export async function start(): Promise<void> {
     serverApp = new App()
 
-    const port = parseInt(process.env.PORT || '', 10) || 3333
+    const port = parseInt(process.env.PORT || '', 10) || 3000
     const server = http.createServer(serverApp.app)
 
     const io = new Server(server, {
